File: shared/src/test/scala/org/pac4j/lagom/scaladsl/test/DefaultClientTest.scala
package org.pac4j.lagom.scaladsl.test

import com.lightbend.lagom.scaladsl.api.transport.{Forbidden, RequestHeader}
import com.lightbend.lagom.scaladsl.server.{LagomApplication, LagomServer, LocalServiceLocator}
import com.lightbend.lagom.scaladsl.testkit.ServiceTest
import org.pac4j.core.context.HttpConstants.AUTHORIZATION_HEADER
import org.pac4j.lagom.scaladsl.{TestModule, TestService, TestServiceImpl}
import org.scalatest.{AsyncWordSpec, BeforeAndAfterAll, Matchers}
import play.api.libs.ws.ahc.AhcWSComponents
import com.lightbend.lagom.scaladsl.testkit.ServiceTest.TestServer
import com.softwaremill.macwire.wire

/**
  * Test of security logic for default client ({@link org.pac4j.http.client.direct.HeaderClient}).
  *
  * @author Vladimir Kornyshev
  * @since 1.0.0
  */
class DefaultClientTest extends AsyncWordSpec with Matchers with BeforeAndAfterAll {

  lazy val server: TestServer[LagomApplication with LocalServiceLocator with AhcWSComponents] = {
    ServiceTest.startServer(ServiceTest.defaultSetup.withCluster(false)) { ctx =>
      new LagomApplication(ctx) with LocalServiceLocator with AhcWSComponents with TestModule {
        override def lagomServer: LagomServer = LagomServer.forService(
          bindService[TestService].to(wire[TestServiceImpl])
        )
      }
    }
  }

  lazy val service: TestService = server.serviceClient.implement[TestService]

  "TestService" should {

    "authenticate by anonymous" in {
      service.defaultAuthenticate.invoke.map { result =>
        result should ===("anonymous")
      }
    }

    "authenticate by Alice" in {
      service.defaultAuthenticate.handleRequestHeader((header: RequestHeader) => header.withHeader(AUTHORIZATION_HEADER, "Alice")).invoke.map { result =>
        result should ===("Alice")
      }
    }

    "not authorize by anonymous" in {
      service.defaultAuthorize.invoke.map { result =>
        fail("authorize by anonymous should be forbidden")
      } recoverWith {
        case f: Forbidden =>
          f.getMessage should ===("Authorization failed")
      }
    }

    "authorize by Alice" in {
      service.defaultAuthorize.handleRequestHeader((header: RequestHeader) => header.withHeader(AUTHORIZATION_HEADER, "Alice")).invoke.map { result =>
        result should ===("Alice")
      }
    }

    "not authorize by anonymous (authorizer from config)" in {
      service.defaultAuthorizeConfig.invoke.map { result =>
        fail("authorize by anonymous should be forbidden")
      } recoverWith {
        case f: Forbidden =>
          f.getMessage should ===("Authorization failed")
      }
    }

    "authorize by Alice (authorizer from config)" in {
      service.defaultAuthorizeConfig.handleRequestHeader((header: RequestHeader) => header.withHeader(AUTHORIZATION_HEADER, "Alice")).invoke.map { result =>
        result should ===("Alice")
      }
    }

  }

  override protected def beforeAll(): Unit = server

  override protected def afterAll(): Unit = server.stop()

}

File: shared/src/test/java/org/pac4j/lagom/javadsl/test/DefaultClientTest.java
package org.pac4j.lagom.javadsl.test;

import com.lightbend.lagom.javadsl.api.transport.Forbidden;
import com.lightbend.lagom.javadsl.testkit.ServiceTest.TestServer;
import org.junit.jupiter.api.AfterAll;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;
import org.pac4j.lagom.javadsl.TestService;

import java.util.concurrent.ExecutionException;

import static com.lightbend.lagom.javadsl.testkit.ServiceTest.defaultSetup;
import static com.lightbend.lagom.javadsl.testkit.ServiceTest.startServer;
import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.catchThrowable;
import static org.pac4j.core.context.HttpConstants.AUTHORIZATION_HEADER;

/**
 * Test of security logic for default client ({@link org.pac4j.http.client.direct.HeaderClient}).
 *
 * @author Sergey Morgunov
 * @since 1.0.0
 */
class DefaultClientTest {

    private static TestServer server;

    private static TestService service;

    @BeforeAll
    static void beforeAll() {
        server = startServer(defaultSetup().withCluster(false));
        service = server.client(TestService.class);
    }

    @AfterAll
    static void afterAll() {
        if (server != null) server.stop();
    }

    @Test
    @DisplayName("authenticate by anonymous")
    void testAuthenticateAnonymous() throws ExecutionException, InterruptedException {
        String result = service.defaultAuthenticate().invoke().toCompletableFuture().get();
        assertThat(result).isEqualTo("anonymous");
    }

    @Test
    @DisplayName("authenticate by Alice")
    void testAuthenticateProfile() throws ExecutionException, InterruptedException {
        String result = service.defaultAuthenticate()
                .handleRequestHeader(header -> header.withHeader(AUTHORIZATION_HEADER, "Alice"))
                .invoke().toCompletableFuture().get();
        assertThat(result).isEqualTo("Alice");
    }

    @Test
    @DisplayName("authorize by anonymous")
    void testAuthorizeAnonymous() {
        Throwable thrown = catchThrowable(() -> service.defaultAuthorize().invoke().toCompletableFuture().get());
        assertThat(thrown).hasCauseExactlyInstanceOf(Forbidden.class);
        assertThat(thrown.getCause()).hasMessage("Authorization failed");
    }

    @Test
    @DisplayName("authorize by Alice")
    void testAuthorizeProfile() throws ExecutionException, InterruptedException {
        String result = service.defaultAuthorize()
                .handleRequestHeader(header -> header.withHeader(AUTHORIZATION_HEADER, "Alice"))
                .invoke().toCompletableFuture().get();
        assertThat(result).isEqualTo("Alice");
    }

    @Test
    @DisplayName("authorize by anonymous (authorizer from config)")
    void testAuthorizeConfigAnonymous() {
        Throwable thrown = catchThrowable(() -> service.defaultAuthorizeConfig().invoke().toCompletableFuture().get());
        assertThat(thrown).hasCauseExactlyInstanceOf(Forbidden.class);
        assertThat(thrown.getCause()).hasMessage("Authorization failed");
    }

    @Test
    @DisplayName("authorize by Alice (authorizer from config)")
    void testAuthorizeConfigProfile() throws ExecutionException, InterruptedException {
        String result = service.defaultAuthorizeConfig()
                .handleRequestHeader(header -> header.withHeader(AUTHORIZATION_HEADER, "Alice"))
                .invoke().toCompletableFuture().get();
        assertThat(result).isEqualTo("Alice");
    }

}
